src/boot/bootctl-install.c

```diff
@@ -334,9 +338,16 @@ static int update_efi_boot_binaries(const char *esp_path, const char *source_pat
                 _cleanup_close_ int fd = -EBADF;
                 _cleanup_free_ char *v = NULL;
 
+                if (de->d_type != DT_REG)
+                        continue;
+
                 if (!endswith_no_case(de->d_name, ".efi"))
                         continue;
 
+                /* Skip the file that is already updated. */
+                if (path_equal_filename(de->d_name, exclude_path))
```

Comment:
This doesn't account for the case insensitivity of the file system. That's what caused [the NixOS bug](https://github.com/NixOS/nixpkgs/issues/331170) that forced us not to use this PR.

```suggestion
                if (inode_same_at(dirfd(d), de->d_name, 0, exclude_path, 0))
```

This seems to work, and we may be start shipping this in NixOS soon so we no longer have to revert f8b02a56febf14adf2474875a1b6625f1f346a6f